### PR TITLE
feat(providers) initial implementation of pulumi provider

### PR DESCRIPTION
**Pull request by agershman**
This is an initial implementation of a Pulumi provider backed by state. This first cut meets my use case of obtaining state from the Pulumi REST API. I've updated the readme with the relevant usage and a couple examples.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -234,9 +236,9 @@ Please see [pkg/providers](https://github.com/helmfile/vals/tree/master/pkg/prov
 
 The `auth_method` or `VAULT_AUTH_METHOD` envar configures how `vals` authenticates to HashiCorp Vault. Currently only these options are supported:
 
-* [approle](https://www.vaultproject.io/docs/auth/approle#via-the-api): it requires you pass on a `role_id` together with a `secret_id`.
-* [token](https://www.vaultproject.io/docs/auth/token): you just need creating and passing on a `VAULT_TOKEN`. If `VAULT_TOKEN` isn't set, token can be retrieved from `VAULT_TOKEN_FILE` env or `~/.vault-token` file.
-* [kubernetes](https://www.vaultproject.io/docs/auth/kubernetes): if you're running inside a Kubernetes cluster, you can use this option. It requires you [configure](https://www.vaultproject.io/docs/auth/kubernetes#configuration) a policy, a Kubernetes role, a service account and a JWT token. The login path can also be set using the environment variable `VAULT_KUBERNETES_MOUNT_POINT` (default is `/kubernetes`). You must also set `role_id` or `VAULT_ROLE_ID` envar to the Kubernetes role.
+- [approle](https://www.vaultproject.io/docs/auth/approle#via-the-api): it requires you pass on a `role_id` together with a `secret_id`.
```
Comment (yxxhero) [inline README.md]: why change?

Comment (agershman) [reply README.md]: Looks like my vscode markdown formatter went to town. I'll fix and amend today.

Comment (agershman) [reply README.md]: @yxxhero I've amended with relevant `README.md` changes only.

**Comment by yxxhero** [comment]
@agershman please fix ci issues.

**Comment by agershman** [comment]
> @agershman please fix ci issues.

Should be good to go now.

**Comment by agershman** [comment]
> @agershman please fix ci issues.

@yxxhero should be good to go now. Thanks

File: pkg/providers/pulumi/pulumi.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+	p.stack = cfg.String("stack")
+
+	p.log.Debugf("pulumi: backend=%q, api_endpoint=%q, organization=%q, project=%q, stack=%q",
+		p.backend, p.pulumiAPIEndpointURL, p.organization, p.project, p.stack)
+
+	return p
+}
+
+func (p *provider) GetString(key string) (string, error) {
+	tokens := strings.Split(key, "/")
```
Comment (yxxhero) [inline pkg/providers/pulumi/pulumi.go:101]: @agershman we should check the length of tokens here. WDYT?

File: pkg/providers/pulumi/pulumi.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+	// https://gjson.dev/
+	resourceAttributePath := tokens[3]
+
+	var state *pulumiState
+	var err error
+	switch p.backend {
+	case "pulumistateapi":
+		state, err = p.getStateFromPulumiAPI()
+	default:
+		return "", fmt.Errorf("unsupported backend: %s", p.backend)
```
Comment (yxxhero) [inline pkg/providers/pulumi/pulumi.go]: err = fmt.Errorf("unsupported backend: %s", p.backend)

Comment (agershman) [reply pkg/providers/pulumi/pulumi.go]: Will address.

File: pkg/providers/pulumi/pulumi.go (hunk 2 above)
Comment (agershman) [reply pkg/providers/pulumi/pulumi.go:101]: Yep, I think that'd be helpful to be more explicit if the provided key is not in a supported format. Will address.

**Comment by agershman** [comment]
@yxxhero I'd like to get the pulumi vals support in place on the helmfile side, which I believe would require a release on the vals project. It looks like releases aren't auto created on PR merge and so was wondering if you could kick that off?

**Pull request merged**